### PR TITLE
issue #22762 fixed - Credit Memo with Zero Total: Order Status "Complete" and not "Closed"

### LINKED ISSUE #22762: Credit Memo with Zero Total: Order Status "Complete" and not "Closed"
### Preconditions (*)

1. Magento 2.3.1
2. PHP7.2

### Steps to reproduce (*)

1. Create order with zero Amount (all Grand Total, Total Paid, Total Refunded, Total Due, Subtotal, Shipping & Handling, Tax are zero ) in status = "Pending" and 1 SKU qty = 1.
2. Create Invoice and now Order is status = "Processing"
3. Ship Order -> now Orderstatus is = "Complete"
4. Now create Credit memo and refund and return product

### Expected result (*)

1. The order status should be "Closed"

### Actual result (*)

1. Order Status is "Complete"

### Additional information
- [Order Status Workflow](https://docs.magento.com/user-guide/sales/order-status-workflow.html)

### PR DESCRIPTION
**Pull request by ghost**
### Description (*)
I have added a function to check for credit memo in Order.php, I had to edit StateResolver.php because the status in notification was showing complete.

### Fixed Issues (if relevant)
1. magento/magento2#22762: Credit Memo with Zero Total: Order Status "Complete" and not "Closed"

### Manual testing scenarios (*)
1. Created a product with 0 price 
2. Ordered with free shipping
3. Created Invoice then Credit memo, now the Order status is "Closed"
4. Created a product with non zero price applied 100% discount coupon.
5. Followed step 2 and 3.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @sanjaychouhan-webkul,
Thank you for your contribution!

It looks like that your PR fix the issue in most cases, but there are still cases where it does not work properly.

Could you please check next case?
**Steps to reproduce:**
1. Create Order with zero total with 2 products.
2. Create Invoice and Shipment.
3. Do partial credit memo for one of the products.
4. Check order state

**Expected**
1. Order still in status 'Complete', as there are products, which still not refunded.
2. You are able to create second partial refund.

**Actual**
1. Order in status 'Closed'.
2. Second product can not be refunded.

image

**Comment by ghost** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @sanjaychouhan-webkul. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @sanjaychouhan-webkul, here is your [Magento instance](https://i-22825-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-22825-2-3-develop.instances.magento-community.engineering/admin](https://i-22825-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
I have another issue.

**Steps to reproduce:**
1. If we order two products (one with 0 price and other with non zero) 
2. Create invoice for both.
3. Now create credit memo for non zero priced product.

**Expected**
1. Credit Memo button should appear because we have not created credit memo for the other product.

**Actual**
1. Credit memo button does not appear.
Screenshot_534

**Comment by ghost** [comment]
I am working on fixing all the issues in next commit

**Comment by ghost** [comment]
@swnsma I have fixed all related issues. Please check.

File: app/code/Magento/Sales/Model/Order/StateResolver.php
Comment (swnsma) [inline app/code/Magento/Sales/Model/Order/StateResolver.php]: Could you please check cases when we really need this check `0 == $order->getBaseGrandTotal()` ?
Probably, we can much simplify and leave only `$order->canCreditmemo()` check.
It will much simplify the solution.

File: app/code/Magento/Sales/Model/Order.php
Comment (swnsma) [inline app/code/Magento/Sales/Model/Order.php:697]: Do we need to change something here? 
It seams call of this method appears only when order has `grand total > 0`.

File: app/code/Magento/Sales/Model/Order.php
Comment (swnsma) [inline app/code/Magento/Sales/Model/Order.php]: Does it make sense? 
It seams that `getTotalDue` for zero order will be always `0`.

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @sanjaychouhan-webkul,

Thank you for your changes.
I think we can much simplify current solution. 
Could we start with check that `0 == $order->getBaseGrandTotal()` is really required in `\Magento\Sales\Model\Order\StateResolver::isOrderComplete`

Also, please check failed static tests.

File: app/code/Magento/Sales/Model/Order.php
Comment (ghost) [reply app/code/Magento/Sales/Model/Order.php:697]: This is to fix the issue which I mentioned earlier with (1 non zero product) https://github.com/magento/magento2/pull/22825#issuecomment-492124318

File: app/code/Magento/Sales/Model/Order.php
Comment (ghost) [reply app/code/Magento/Sales/Model/Order.php]: I think we should remove the check for $paidAmtIsRefunded because it return **true** so the function **canCreditmemoForZeroTotal** returns false and so we are unable to generate credit memo for second time in the below scenario
1. We order 2 quantities of zero priced product 
2. Create invoice for both the products
3. Generate credit-memo for one product.
Now the Credit memo button will not appear, but it should because we have not refunded all the products.

I think we should remove $paidAmtIsRefunded, **$this->isAllInvoicedRefunded()** will solve the problem to check if all refunded. 

@swnsma  Please let me know.

File: app/code/Magento/Sales/Model/Order/StateResolver.php
Comment (ghost) [reply app/code/Magento/Sales/Model/Order/StateResolver.php]: I am not sure about the `0 == $order->getBaseGrandTotal()` that's why I didn't remove it.
@orlangur can you suggest what should we do?

**Comment by ghost** [comment]
@swnsma I have applied the changes. Please check.

File: app/code/Magento/Sales/Model/Order/StateResolver.php
Comment (swnsma) [reply app/code/Magento/Sales/Model/Order/StateResolver.php]: Hi @sanjaychouhan-webkul,

Can we just remove this line and check if Integration tests will fail?
Probably, this is only 'hack' was done for zero-total orders.

It seems really strange, because it is difficult to imagine situation where `0 == $order->getBaseGrandTotal()` will be `false` and 
https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/app/code/Magento/Sales/Model/Order.php#L667
will be `true`.

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @sanjaychouhan-webkul,

I am appreciate your work.
Could we start with deletion of `0 == $order->getBaseGrandTotal()` and running of Integration tests?

**Comment by ghost** [comment]
> Hi @sanjaychouhan-webkul,
> 
> I am appreciate your work.
> Could we start with deletion of `0 == $order->getBaseGrandTotal()` and running of Integration tests?

I have removed it in my last commit and Integration Test was successfull. Please check. @swnsma 
https://github.com/magento/magento2/blob/ff2a9fbc7c300ab6257c27f8b84a89fb35cd11da/app/code/Magento/Sales/Model/Order/StateResolver.php#L22-L25

File: app/code/Magento/Sales/Model/Order/StateResolver.php
Comment (ghost) [reply app/code/Magento/Sales/Model/Order/StateResolver.php]: I have removed it in my last commit and Integration Test was successfull. Please check. @swnsma 
https://github.com/magento/magento2/blob/ff2a9fbc7c300ab6257c27f8b84a89fb35cd11da/app/code/Magento/Sales/Model/Order/StateResolver.php#L22-L25

**Comment by ghost** [comment]
@swnsma @VladimirZaets Please check the latest commits

**Comment by swnsma** [comment]
Hi @sanjaychouhan-webkul,
Sorry for long delay. I did not forget about your PR.
I am currently working for creating map of conditions for Zero/not Zero Credit Memo creation.
I will check and update you when finish.

**Comment by swnsma** [comment]
Hi @sanjaychouhan-webkul,

I need to discuss your PR and issue in general with Magento Architect Team.
I will come back to review as soon as all requirements will be clear.

**Comment by ghost** [comment]
> Hi @sanjaychouhan-webkul,
> 
> I need to discuss your PR and issue in general with Magento Architect Team.
> I will come back to review as soon as all requirements will be clear.

**Comment by ghost** [comment]
@swnsma Thank you for the notification. 
I had accidentally closed the pr.
Anyway please let me know when you have any information regarding it.

**Comment by ghost** [comment]
@swnsma what is the status?

**Comment by swnsma** [comment]
Hi @sanjaychouhan-webkul,
I am still waiting when @akaplya come back from vacation and review our case.

File: app/code/Magento/Sales/Model/Order.php
Comment (marshallthornton) [inline app/code/Magento/Sales/Model/Order.php:698]: Should this be false instead of true? If no credit memo collection was returned that should mean there is currently no credit memo attached to the order and it is not refunded.

**Comment by ihor-sviziev** [comment]
@sdzhepa @slavvka,
Is there any updates on this PR? was it approved?
This PR got stuck for half a year

**Comment by swnsma** [comment]
@ihor-sviziev,
I have created [proposal to architecture](https://github.com/magento/architecture/pull/194) to figure out correct 'Can Credit Memo' flow and it still waiting for review by @joni-jones

**Comment by sanjaychouhan-webkul** [comment]
Any update on this?

**Comment by swnsma** [comment]
Hi @sanjaychouhan-webkul!
It was a too-long road to the decision.
Could you please align 'can credit memo' behavior with the diagram?
https://github.com/magento/architecture/pull/194

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @sanjaychouhan-webkul!
It was a too-long road to the decision.
Could you please align 'can credit memo' behavior with the diagram?
magento/architecture#194

**Comment by swnsma** [comment]
Hi @sanjaychouhan-webkul,
Are you going to continue with this PR?

**Comment by engcom-Charlie** [comment]
Hi @sanjaychouhan-webkul, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Pull request closed without merging**
